sda-commons-web-autoconfigure/src/main/java/org/sdase/commons/spring/boot/web/security/handler/ObscuringErrorHandler.java

```diff
@@ -0,0 +1,92 @@
+import org.springframework.http.server.ServerHttpResponse;
+import org.springframework.http.server.ServletServerHttpResponse;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;
+
+/**
+ * Error handle that replaces default errors responses with custom {@link ApiError}.
+ *
+ * <p>This handler is invoked for explicitly defined error responses like {@code return
+ * ResponseEntity.status(404).build();} and is not invoked for response indirectly create by {@code
+ * throw new NotFoundException()}.
+ *
+ * <p>This handler addresses risks identified in the security guide as:
+ *
+ * <ul>
+ *   <li>"Risk: Detection of confidential components ... Removal of application related Error
+ *       messages."
+ * </ul>
+ */
+@RestControllerAdvice
+public class ObscuringErrorHandler implements ResponseBodyAdvice<Object> {
+  private static final String RESPONSE_ENTITY_TYPE = ResponseEntity.class.getName();
+  private static final String ACTUATOR_PACKAGE_NAME =
+      "org.springframework.boot.actuate.endpoint.web.servlet";
+  private static final String API_ERROR_RESPONSE_ENTITY_TYPE =
+      String.format("%s<%s>", ResponseEntity.class.getTypeName(), ApiError.class.getTypeName());
+
+  @Override
+  public boolean supports(
+      MethodParameter returnType, Class<? extends HttpMessageConverter<?>> converterType) {
+    return shouldTransformResponse(returnType);
+  }
+
+  @Override
+  public Object beforeBodyWrite(
+      Object body,
+      MethodParameter returnType,
+      MediaType selectedContentType,
+      Class<? extends HttpMessageConverter<?>> selectedConverterType,
+      ServerHttpRequest request,
+      ServerHttpResponse response) {
+    var responseStatus =
+        HttpStatus.resolve(((ServletServerHttpResponse) response).getServletResponse().getStatus());
+
+    // No replacement for `ResponseEntity.status(204).build()`
+    if (responseStatus == null || !responseStatus.isError()) {
+      return body;
+    }
+
+    // Replacement error responses with standard errors, e.g
+    // `ResponseEntity.internalServerError().body(customErrorBody)
+    var apiError = new ApiError();
+    apiError.setTitle("HTTP Error " + responseStatus.value() + " occurred.");
+    return apiError;
+  }
+
+  private boolean shouldTransformResponse(MethodParameter returnType) {
+    // No replacement for `ResponseEntity<ApiError>`
+    if (API_ERROR_RESPONSE_ENTITY_TYPE.equals(returnType.getGenericParameterType().getTypeName())) {
+      return false;
+    }
+
+    // No replacement for `ResponseEntity<MyDto>`
+    if (!RESPONSE_ENTITY_TYPE.equals(returnType.getParameterType().getTypeName())) {
+      return false;
+    }
+
+    // No replacement for `ResponseEntity<Object>` returned by spring actuator
+    if (ACTUATOR_PACKAGE_NAME.equals(
+        returnType.getExecutable().getDeclaringClass().getPackageName())) {
+      return false;
+    }
```

Comment:
Do you know what will happen in case of `ResponseEntity<Void>`?

Reply:
produces the our ApiError. I [added a test](https://github.com/SDA-SE/sda-spring-boot-commons/pull/129/commits/ca357183ce1f23d6c125ea516ac82abdbb570de7) for that.